superset-frontend/src/views/datasetList/DatasetModal.tsx

```diff
@@ -0,0 +1,318 @@
+      .then(({ json: datasourcesJson = {} }) => {
+        const datasources = datasourcesJson.result.map(
+          ({ text, value }: { text: string; value: number }) => ({
+            label: text,
+            value: `${value}`,
+          }),
+        );
+        this.setState({ datasources });
+      })
+      .catch(e => {
+        this.props.addDangerToast(
+          t('An error occurred while fetching datasources'),
+        );
+        if (e) {
+          console.error(e);
+        }
+      });
+  }
+
+  fetchSchemas(datasourceId: string) {
+    if (datasourceId) {
+      this.setState({ schemaLoading: true });
+      const endpoint = `/superset/schemas/${datasourceId}/false/`;
+      SupersetClient.get({ endpoint })
+        .then(({ json: schemaJson = {} }) => {
+          const schemas = schemaJson.schemas.map((s: string) => ({
+            value: s,
+            label: s,
+          }));
+          this.setState({
+            disableSelectSchema: false,
+            schemaLoading: false,
+            schemas,
+          });
+        })
+        .catch(e => {
+          this.setState({ schemaLoading: false, schemas: [] });
+          this.props.addDangerToast(t('Error while fetching schema list'));
+          if (e) {
+            console.error(e);
+          }
+        });
+    }
+  }
+
+  fetchTables(schema: string) {
+    const { datasource } = this.state;
+
+    if (datasource && schema) {
+      this.setState(() => ({ tableLoading: true, tables: [] }));
+      const endpoint = encodeURI(
+        `/superset/tables/${datasource.value}/` +
+          `${encodeURIComponent(schema)}/undefined/false/`,
+      );
+      SupersetClient.get({ endpoint })
+        .then(({ json: tableJson = {} }) => {
+          const options = tableJson.options.map((o: option) => ({
+            value: o.value,
+            label: o.label,
+          }));
+          this.setState(() => ({
+            disableSelectTable: false,
+            tableLoading: false,
+            tables: options,
+          }));
+        })
+        .catch(e => {
+          this.setState(() => ({ tableLoading: false, tables: [] }));
+          this.props.addDangerToast(t('Error while fetching table list'));
+          if (e) {
+            console.error(e);
+          }
+        });
+    }
+  }
+
+  renderFields() {
+    const {
+      datasource,
+      datasources,
+      disableSelectSchema,
+      disableSelectTable,
+      schema,
+      schemaLoading,
+      schemas,
+      table,
+      tableLoading,
+      tables,
+    } = this.state;
+    const { onDatabaseChange, onSchemaChange, onTableChange } = this;
+
+    return (
+      <>
+        <FieldTitle>{t('datasource')}</FieldTitle>
+        <Select
+          clearable={false}
+          ignoreAccents={false}
+          name="select-datasource"
+          onChange={onDatabaseChange}
+          options={datasources}
+          placeholder={t('Select Datasource')}
+          value={datasource}
+          width={500}
```

Comment:
can you pull this out into a reusable constant for all the Select components here?

or possibly even better, create a const at the top of the file for all the shared select params (`clearable`, `ignoreAccents`, `width`) and then object spread it for all the components here